### PR TITLE
Refactoring: remove CompltetionItemFilter

### PR DESCRIPTION
**Pull request by ivanbasov**

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs [hunk 1]
```diff
@@ -7,11 +7,13 @@
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Completion.Providers;
 using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Completion.CompletionProviders;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion;
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs]: ❔ Was this one necessary?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ReferenceDirectiveCompletionProviderTests.cs [hunk 2]
```diff
@@ -7,8 +7,10 @@
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Completion;
 using Microsoft.CodeAnalysis.Editor.CSharp.Completion.FileSystem;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion;
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ReferenceDirectiveCompletionProviderTests.cs]: ❔ Was this one necessary?

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Tags;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using Microsoft.VisualStudio.Text.Adornments;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+    internal static class CompletionItemFilter
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs]: 😕 I was expecting this to be removed

Also, do you know why it didn't show up as a rename in source control? Can you trying removing all unnecessary edits relative to the original and amend the commit so Git detects it as a rename instead of unrelated files?

**Comment by sharwell** [comment]
@ivanbasov Can you give a high level overview of the goal of this pull request? I got lost in the middle and wasn't sure what the original expectations were.

**Comment by ivanbasov** [comment]
@sharwell this is a continuation of removing the old completion. The class is used internally only and there is an unnecessary conversion to the editor class and a cache which can be simplified.

**Comment by genlu** [comment]
Maybe do this in master-vs-deps directly? I think I have made changes in CompletionSource that might cause a lot of conflict with this. Or just wait a few more days, since my change would be merged back to master from 16.3-vs-deps once 16.3 GA :)

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            ConstantFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Constants, WellKnownTags.Constant, 'o', builder);
+            FieldFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Fields, WellKnownTags.Field, 'f', builder);
+            EventFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Events, WellKnownTags.Event, 'v', builder);
+            PropertyFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Properties, WellKnownTags.Property, 'p', builder);
+            MethodFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Methods, WellKnownTags.Method, 'm', builder);
+            ExtensionMethodFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Extension_methods, WellKnownTags.ExtensionMethod, 'x', builder);
+            LocalAndParameterFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Locals_and_parameters, ImmutableArray.Create(WellKnownTags.Local, WellKnownTags.Parameter), 'l', builder);
+            KeywordFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Keywords, ImmutableArray.Create(WellKnownTags.Keyword), 'k', builder);
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs]: nit: I think the last three calls can use another `CreateCompletionFilterAndAddToBuilder` overload

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        private static CompletionFilter CreateCompletionFilterAndAddToBuilder(
+            string displayText, string tag, char accessKey,
+            ImmutableDictionary<string, CompletionFilter>.Builder builder)
+            => CreateCompletionFilterAndAddToBuilder(displayText, ImmutableArray.Create(tag), accessKey, builder);
+
+        public static ImmutableArray<CompletionFilter> GetFilters(RoslynCompletionItem item)
+        {
+            var result = ArrayBuilder<CompletionFilter>.GetInstance();
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemFilter.cs]: creating/freeing `ArrayBuilder` can now be done using using statement, just in case you are not aware of this change :)

e.g.
[http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/InlineDeclaration/CSharpInlineDeclarationCodeFixProvider.cs,76](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/InlineDeclaration/CSharpInlineDeclarationCodeFixProvider.cs,76)

**Comment by CyrusNajmabadi** [comment]
There's no description of this.  Can a synopsis of the thinkin/decision be provided here?  Thanks!

**Comment by CyrusNajmabadi** [comment]
Oh.  this doesn't esem like a remove.  More of a 'move'.

In that case, consider two PRs.  one that actually moves the file, the other that renames the symbols.

**Comment by ivanbasov** [comment]
> Also, have you confirmed this move wouldn't break things like VSMac or VSCode that's hosted in a different editor? For example, Mono has IVT to Ms.CA.Features

Thanks, @genlu ! It is interesting. I hope that internal classes should not be consumed by anyone else but Roslyn. But let us confirm with VSMac? @KirillOsenkov , could you please double-check that VS Mac does not use Roslyn CompletionItemFilter?

**Comment by KirillOsenkov** [comment]
@sandyarmstrong

**Comment by sandyarmstrong** [comment]
@ivanbasov @genlu @KirillOsenkov VSmac only uses it in the old editor, which uses legacy completion API. Of course, we have to remove this when we adopt Roslyn 3.4, so it shouldn't be a problem.

**Comment by ivanbasov** [comment]
@genlu I merged with code you merged from dev16.3-vs-deps. Then, I had to make some changes. Please take a look once again.

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            MethodFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Methods, 'm', WellKnownTags.Method);
+            ExtensionMethodFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Extension_methods, 'x', WellKnownTags.ExtensionMethod);
+            LocalAndParameterFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Locals_and_parameters, 'l', WellKnownTags.Local, WellKnownTags.Parameter);
+            KeywordFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Keywords, 'k', WellKnownTags.Keyword);
+            SnippetFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Snippets, 't', WellKnownTags.Snippet);
+            TargetTypedFilter = CreateCompletionFilterAndAddToBuilder(FeaturesResources.Target_type_matches, 'j', WellKnownTags.TargetTypeMatch);
+
+            s_filterMap = builder.ToImmutable();
+
+            s_expanderMask = BitVector32.CreateMask(previousMask);
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs]: When I implemented this, I found it counter intuitive that `CreateMask` do the shift for you. Then surely, I was confused by this again today :D

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+
+    /// <summary>
+    /// Provides an efficient way to compute a set of completion filters associated with a collection of completion items.
+    /// Presence of expander and filter in the set have different meanings. Set contains a filter means the filter is
+    /// available but unselected, whereas it means available and selected for an expander. Note that even though VS supports 
+    /// having multiple expanders, we only support one.
+    /// </summary>
+    internal class FilterSet
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs]: nit: sealed

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                foreach (var tag in tags)
+                {
+                    builder.Add(tag, new FilterWithMask(filter, previousMask));
+                }
+
+                return filter;
+            }
+        }
+
+        public static readonly CompletionFilter NamespaceFilter;
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterSet.cs]: nit: maybe move all fields to the top?

**Comment by genlu** [comment]
Reviewed again. LGTM!
